src/editor.tsx

```diff
@@ -1,186 +1,188 @@
-
-  constructor(props: MonacoEditorProps) {
-    super(props);
-    this.containerElement = undefined;
-  }
-
-  componentDidMount() {
-    this.initMonaco();
-  }
-
-  componentDidUpdate(prevProps: MonacoEditorProps) {
-    const { value, language, theme, height, options, width, className } =
-      this.props;
-
-    const { editor } = this;
-    const model = editor.getModel();
-
-    if (this.props.value != null && this.props.value !== model.getValue()) {
-      this.__prevent_trigger_change_event = true;
-      this.editor.pushUndoStop();
+function MonacoEditor({
+  width,
+  height,
+  value,
+  defaultValue,
+  language,
+  theme,
+  options,
+  overrideServices,
+  editorWillMount,
+  editorDidMount,
+  editorWillUnmount,
+  onChange,
+  className,
+}: MonacoEditorProps) {
+  const containerElement = useRef<HTMLDivElement | null>(null);
+
+  const editor = useRef<monaco.editor.IStandaloneCodeEditor | null>(null);
+
+  const _subscription = useRef<monaco.IDisposable | null>(null);
+
+  const __prevent_trigger_change_event = useRef<boolean | null>(null);
+
+  const fixedWidth = processSize(width);
+
+  const fixedHeight = processSize(height);
+
+  const style = useMemo(
+    () => ({
+      width: fixedWidth,
+      height: fixedHeight,
+    }),
+    [fixedWidth, fixedHeight]
+  );
+
+  const handleEditorWillMount = useCallback(() => {
+    const finalOptions = editorWillMount(monaco);
+    return finalOptions || {};
+  }, [editorWillMount]);
+
+  const handleEditorDidMount = useCallback(() => {
+    editorDidMount(editor.current, monaco);
+
+    _subscription.current = editor.current.onDidChangeModelContent((event) => {
+      if (!__prevent_trigger_change_event.current) {
+        onChange(editor.current.getValue(), event);
+      }
+    });
+  }, [editorDidMount, onChange]);
+
+  const handleEditorWillUnmount = useCallback(() => {
+    editorWillUnmount(editor.current, monaco);
+  }, [editorWillUnmount]);
+
+  const initMonaco = useCallback(() => {
+    const finalValue = value !== null ? value : defaultValue;
+    if (containerElement.current) {
+      // Before initializing monaco editor
+      const finalOptions = { ...options, ...handleEditorWillMount() };
+      editor.current = monaco.editor.create(
+        containerElement.current,
+        {
+          value: finalValue,
+          language,
+          ...(className ? { extraEditorClassName: className } : {}),
+          ...finalOptions,
+          ...(theme ? { theme } : {}),
+        },
+        overrideServices
+      );
+      // After initializing monaco editor
+      handleEditorDidMount();
+    }
+  }, [
+    className,
+    defaultValue,
+    handleEditorDidMount,
+    handleEditorWillMount,
+    language,
+    options,
+    overrideServices,
+    theme,
+    value,
+  ]);
+
+  useEffect(() => {
+    initMonaco();
+  }, [initMonaco]);
+
+  useEffect(() => {
+    if (editor.current) {
+      const model = editor.current.getModel();
```

Comment:
`getModel()` can return `null`. This is causing our tests to fail.

https://github.com/pybricks/pybricks-code/actions/runs/3576583892/jobs/6014566670